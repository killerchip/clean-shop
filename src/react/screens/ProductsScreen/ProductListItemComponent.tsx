import { ProductListItem } from "@/presentation/products.views";
import { Card, MD3Theme, withTheme } from "react-native-paper";
import { Pressable, StyleProp, ViewStyle } from "react-native";
import { useRouter } from "expo-router";
import { PriceTag } from "../../components/atoms/PriceTags";
import styled from "styled-components/native";

type Props = {
  product: ProductListItem;
  style?: StyleProp<ViewStyle>;
  testID?: string;
};

export const ProductListItemComponent = function ProductListItem({
  product,
  style = {},
  testID,
}: Props) {
  const router = useRouter();

  const onPress = () => {
    router.push({ pathname: `/product/${product.id}` });
  };

  return (
    <Pressable onPress={onPress} testID={`${testID}-pressable-area`}>
      <ListItemCard style={style}>
        <Card.Cover source={{ uri: product.image }} />
        <Card.Title title={product.title} titleVariant="titleLarge" />
        <ListItemContent>
          <PriceTag>${product.price}</PriceTag>
        </ListItemContent>
      </ListItemCard>
    </Pressable>
  );
};

const ListItemCard = withTheme(styled(Card)<{ theme: MD3Theme }>`
  margin: 4px;
`);

const ListItemContent = withTheme(styled(Card.Content)<{ theme: MD3Theme }>`
  flex-direction: row;
  justify-content: flex-end;
`);
