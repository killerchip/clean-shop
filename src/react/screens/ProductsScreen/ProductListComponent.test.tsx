import { render, fireEvent, screen } from "@testing-library/react-native";
import { ProductListItemComponent } from "./ProductListItemComponent";
import { useRouter } from "expo-router";

jest.mock("expo-router");

describe("ProductListItemComponent", () => {
  const mockProduct = {
    id: 1,
    title: "Test Product",
    description: "Test Description",
    price: 100,
    image: "test-image-url",
  };

  let mockPush: jest.Mock;

  beforeEach(() => {
    mockPush = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({
      push: mockPush,
    });
  });

  it("renders product details correctly", () => {
    render(<ProductListItemComponent product={mockProduct} />);

    expect(screen.getByText(mockProduct.title)).toBeTruthy();
    expect(screen.getByText(`$${mockProduct.price}`)).toBeTruthy();
  });

  it("matches snapshot", () => {
    render(<ProductListItemComponent product={mockProduct} />);

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("navigates to product detail page on press", () => {
    render(<ProductListItemComponent product={mockProduct} testID="item" />);

    fireEvent.press(screen.getByTestId("item-pressable-area"));
    expect(mockPush).toHaveBeenCalledWith({
      pathname: `/product/${mockProduct.id}`,
    });
  });
});
