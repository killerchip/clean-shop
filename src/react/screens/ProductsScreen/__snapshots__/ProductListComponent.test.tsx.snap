// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListItemComponent matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="undefined-pressable-area"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(247, 243, 249, 1)",
        "borderRadius": 12,
        "marginBottom": 4,
        "marginLeft": 4,
        "marginRight": 4,
        "marginTop": 4,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 3,
      }
    }
    testID="card-container-outer-layer"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(247, 243, 249, 1)",
          "borderRadius": 12,
          "elevation": 1,
          "flex": undefined,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 1,
        }
      }
      testID="card-container"
    >
      <View
        style={
          [
            {
              "flexShrink": 1,
            },
            undefined,
          ]
        }
        testID="card"
      >
        <View
          style={
            [
              {
                "backgroundColor": "#eeeeee",
                "height": 195,
                "overflow": "hidden",
              },
              {
                "borderRadius": 12,
              },
              undefined,
            ]
          }
        >
          <Image
            accessibilityIgnoresInvertColors={true}
            borderRadiusStyles={{}}
            siblings={
              [
                "Card.Cover",
                "Card.Title",
                "withTheme(Styled(Card.Content))",
              ]
            }
            source={
              {
                "uri": "test-image-url",
              }
            }
            style={
              [
                {
                  "flex": 1,
                  "height": undefined,
                  "justifyContent": "flex-end",
                  "padding": 16,
                  "width": undefined,
                },
                {
                  "borderRadius": 12,
                },
              ]
            }
          />
        </View>
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingLeft": 16,
              },
              {
                "minHeight": 50,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 22,
                      "fontWeight": "400",
                      "letterSpacing": 0,
                      "lineHeight": 28,
                    },
                    [
                      {
                        "minHeight": 30,
                        "paddingRight": 16,
                      },
                      {
                        "marginBottom": 2,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
            >
              Test Product
            </Text>
          </View>
          <View />
        </View>
        <View
          borderRadiusStyles={{}}
          style={
            [
              {
                "paddingHorizontal": 16,
              },
              {
                "paddingBottom": 16,
              },
              {
                "flexDirection": "row",
                "justifyContent": "flex-end",
              },
            ]
          }
          theme={
            {
              "animation": {
                "scale": 1,
              },
              "colors": {
                "backdrop": "rgba(50, 47, 55, 0.4)",
                "background": "rgba(255, 251, 254, 1)",
                "elevation": {
                  "level0": "transparent",
                  "level1": "rgb(247, 243, 249)",
                  "level2": "rgb(243, 237, 246)",
                  "level3": "rgb(238, 232, 244)",
                  "level4": "rgb(236, 230, 243)",
                  "level5": "rgb(233, 227, 241)",
                },
                "error": "rgba(179, 38, 30, 1)",
                "errorContainer": "rgba(249, 222, 220, 1)",
                "inverseOnSurface": "rgba(244, 239, 244, 1)",
                "inversePrimary": "rgba(208, 188, 255, 1)",
                "inverseSurface": "rgba(49, 48, 51, 1)",
                "onBackground": "rgba(28, 27, 31, 1)",
                "onError": "rgba(255, 255, 255, 1)",
                "onErrorContainer": "rgba(65, 14, 11, 1)",
                "onPrimary": "rgba(255, 255, 255, 1)",
                "onPrimaryContainer": "rgba(33, 0, 93, 1)",
                "onSecondary": "rgba(255, 255, 255, 1)",
                "onSecondaryContainer": "rgba(29, 25, 43, 1)",
                "onSurface": "rgba(28, 27, 31, 1)",
                "onSurfaceDisabled": "rgba(28, 27, 31, 0.38)",
                "onSurfaceVariant": "rgba(73, 69, 79, 1)",
                "onTertiary": "rgba(255, 255, 255, 1)",
                "onTertiaryContainer": "rgba(49, 17, 29, 1)",
                "outline": "rgba(121, 116, 126, 1)",
                "outlineVariant": "rgba(202, 196, 208, 1)",
                "primary": "rgba(103, 80, 164, 1)",
                "primaryContainer": "rgba(234, 221, 255, 1)",
                "scrim": "rgba(0, 0, 0, 1)",
                "secondary": "rgba(98, 91, 113, 1)",
                "secondaryContainer": "rgba(232, 222, 248, 1)",
                "shadow": "rgba(0, 0, 0, 1)",
                "surface": "rgba(255, 251, 254, 1)",
                "surfaceDisabled": "rgba(28, 27, 31, 0.12)",
                "surfaceVariant": "rgba(231, 224, 236, 1)",
                "tertiary": "rgba(125, 82, 96, 1)",
                "tertiaryContainer": "rgba(255, 216, 228, 1)",
              },
              "dark": false,
              "fonts": {
                "bodyLarge": {
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "letterSpacing": 0.15,
                  "lineHeight": 24,
                },
                "bodyMedium": {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                },
                "bodySmall": {
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "letterSpacing": 0.4,
                  "lineHeight": 16,
                },
                "default": {
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                "displayLarge": {
                  "fontFamily": "System",
                  "fontSize": 57,
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 64,
                },
                "displayMedium": {
                  "fontFamily": "System",
                  "fontSize": 45,
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 52,
                },
                "displaySmall": {
                  "fontFamily": "System",
                  "fontSize": 36,
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 44,
                },
                "headlineLarge": {
                  "fontFamily": "System",
                  "fontSize": 32,
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 40,
                },
                "headlineMedium": {
                  "fontFamily": "System",
                  "fontSize": 28,
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 36,
                },
                "headlineSmall": {
                  "fontFamily": "System",
                  "fontSize": 24,
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 32,
                },
                "labelLarge": {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "letterSpacing": 0.1,
                  "lineHeight": 20,
                },
                "labelMedium": {
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "500",
                  "letterSpacing": 0.5,
                  "lineHeight": 16,
                },
                "labelSmall": {
                  "fontFamily": "System",
                  "fontSize": 11,
                  "fontWeight": "500",
                  "letterSpacing": 0.5,
                  "lineHeight": 16,
                },
                "titleLarge": {
                  "fontFamily": "System",
                  "fontSize": 22,
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 28,
                },
                "titleMedium": {
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "500",
                  "letterSpacing": 0.15,
                  "lineHeight": 24,
                },
                "titleSmall": {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "letterSpacing": 0.1,
                  "lineHeight": 20,
                },
              },
              "isV3": true,
              "roundness": 4,
              "version": 3,
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "rgba(103, 80, 164, 1)",
                },
              ]
            }
          >
            $
            100
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
